### PR TITLE
mon: EINVAL if we get invalid json values for mon commands

### PR DESCRIPTION
**Pull request by liewegas**
http://tracker.ceph.com/issues/25107

### CODE REVIEW

**Comment by batrick** [comment]
@liewegas there's a build failure.

**Comment by badone** [comment]
Jenkins, retest this please.

File: src/common/cmdparse.h [hunk 1]
```diff
@@ -64,11 +64,22 @@ cmd_getval(CephContext *cct, const cmdmap_t& cmdmap, const std::string& k, T& va
...
+cmd_getval(CephContext *cct, const cmdmap_t& cmdmap, const std::string& k,
+	   T& val, const T& defval)
 {
-  if (!cmd_getval(cct, cmdmap, k, val))
+  if (cmdmap.count(k)) {
+    try {
+      val = boost::get<T>(cmdmap.find(k)->second);
+      return true;
+    } catch (boost::bad_get&) {
+      handle_bad_get(cct, k, typeid(T).name());
```
Comment (tchaikov) [inline src/common/cmdparse.h]: this problem also exists in other commands handled in ``OSDMonitor::prepare_command_impl()`.

another way to fix this issue is to rethrow a runtime error with the cmd argument's keyname, and catch this exception in `OSDMonitor::prepare_command_impl()`. so we are able to take care of all commands in a single place once and for all.

**Comment by tchaikov** [comment]
i don't think this PR fixes http://tracker.ceph.com/issues/25107. see http://tracker.ceph.com/projects/ceph/activity?from=2018-07-26 and 
https://github.com/ceph/ceph/pull/23243#discussion_r205323894

**Comment by liewegas** [comment]
merged this one by accident.. anyone want to do a post-whatever review of the exception-based approach?  happy to clean up w/ a follow-on pr

**Pull request merged**